core/templates/dev/head/services/RteHelperService.js

```diff
@@ -181,9 +181,19 @@ oppia.factory('RteHelperService', [
           });
         });
 
-        text = elt.html();
-        text = $filter('removeExtraLines')(text);
-        return text;
+        var textElt = elt[0].childNodes;
+        for (var i = textElt.length - 1; i > 0; i--) {
+          if (textElt[i].localName === 'p') {
+            if (textElt[i].childNodes[0].nodeValue === null ||
```

Comment:
Unless I'm missing something, this feels a bit careless:

- What if there's 0 child nodes?
- What if there's more than one child node?
- What happened to handling `<br>` tags?

Make sure to consider all the possible cases exhaustively.

Reply:
- For the first case, how many child nodes does `<p></p>` have?
- But you never check for "more than one tag", if I'm reading the code correctly.
- What if the paragraph contains a br tag, and then some other stuff?

Reply:
- Here, the for loop is for number of tags in ```<p></p>```. It's the part which contains the text in each line.
- We just have to check for one tag, If there are more than one tags in the paragraph, it means that it doesn't need to check further.
- If there a ```<br>``` then it goes to next line and text is there, if there is some text and new line then it's new para.

Reply:
- I don't think so. textElt[i] represents a "p" node and textElt[i].childNodes represents its children. children[0] takes the first child, but this assumes that there _is_ a first child. That assumption is what I'm asking about. Isn't it possible for the "p" node to contain no children?
- That is not correct either, I think. If the first node is null or just whitespace then your code would remove the paragraph even if the paragraph has later child nodes that aren't just whitespace. That's because the code is examining only the first child node.
- I don't think I understand your response, sorry. I'm talking about the possible case where a paragraph might contain br tags and some other things. This is a special case of point 2. (Also, there might be a case where br tags appear outside paragraphs; your code will not catch this because it is only looking at p tags. I wanted to check whether this is intentional.)

Reply:
- Yes I get this, I've fixed it
- I'm sorry I couldn't generate this possibilty. Attaching pics for reference: 
![screenshot from 2018-03-19 00-12-17](https://user-images.githubusercontent.com/20617436/37571165-0c3b49ea-2b1f-11e8-867a-156f19958fe3.png)
![screenshot from 2018-03-19 02-40-53](https://user-images.githubusercontent.com/20617436/37571166-0c6cc448-2b1f-11e8-84fa-092070213a0b.png)
- Yes that's the technique I'm using for solving the issue. It's expected.

Reply:
Hi @WickedBrat -- re your second/third point, the fix should work in the general case (so assume that any valid HTML string can be produced). This is because we are occasionally changing the framework for the RTE (e.g. there is a GSoC project to move it to CKEditor instead of textAngular) and so this code needs to be future-proof. (So try to analyze and solve the problem cleanly in the general case, even though you may not be able to repro it with the current text editor.)

Also, please make sure to add tests that exercise all the various corner cases!

Reply:
Hi @seanlip I fixed the code as requested, but the frontend tests fail here: 
![screenshot from 2018-03-19 16-01-59](https://user-images.githubusercontent.com/20617436/37590914-c12b6bbc-2b8f-11e8-8647-322b8dbc7fdf.png)

I checked only the 'p' tags for the sane reason.